initial commit

NekhoroshkinaElena opened the pull request.

NekhoroshkinaElena committed:
initial commit

NekhoroshkinaElena committed:
initial commit

NekhoroshkinaElena committed:
initial commit

NekhoroshkinaElena committed:
refactor: убрать множественное дублирование в BookingServiceImpl.java
refactor: добавить метод преобразования в Booking в BookingMapper
checkstyle: иправлять мелкие ошибки

NekhoroshkinaElena committed:
refactor: изменить название методов в BookingController
feat: добавить новый метод преобразования toBookingDtoForItem в класс BookingMapper
refactor: изменить(уточнить сообщения об ошибках)
style: исправлять форматирование

NekhoroshkinaElena commented:
Добрый день! Задание выполнено, столкнулась со множеством трудностей, не уверена, что правильно их разрешила, поэтому очень жду ваших замечаний!:)

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,48 @@
+    private final BookingService bookingService;
+
+    @PostMapping
+    public BookingDtoOutput save(@RequestBody BookingDtoInput bookingDtoInput,
```
Здорово, что назвали здесь DTO с постфиксом Output - верный подход, что используется отдельный объект для того, чтобы отдавать данные

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDtoOutput.java:
```diff
@@ -0,0 +1,30 @@
+
+@Getter
+@Setter
+public class BookingDtoOutput {
```
Кстати, для ДТО есть ещё удобная аннтотация у библиотеки Lombok - @Value аннотация. Позволяет сделать все поля приватными, финальными и добавляет конструктор. 
Тогда можно было бы заменить на такой вариант 
```
@Value
public class BookingDtoOutput {
    long id;
    LocalDateTime start;
    LocalDateTime end;
    BookingStatus status;
    User booker;
    ItemDtoForBooking item;
}
```

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,36 @@
+@RequiredArgsConstructor
+public class BookingMapper {
+
+    public static BookingDtoOutput toBookingDtoOutput(Booking booking) {
+        return new BookingDtoOutput(
```
Отлично, можете ещё посмотреть в сторону библиотеки MapStruct, она сама мапит поля с одинаковым типом и именованием. https://www.baeldung.com/mapstruct

Применить только у вас не получится. Т.к. тесты не пройдут. 
Если будут вопросы как использовать данную аннотацию - можете посмотреть это видео
https://youtu.be/Qczc8ci0ylE?t=409

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -5,18 +5,39 @@
+    @OneToOne
     private User booker;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "status")
```
В целом здесь не обязетельно же ставить аннотацию @Column - hibernate и так создаст столбец с названием status, если бы вы из списка параметров аннотации @Column что-то ещё решили бы задать - тогда можно было бы оставить

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+    public BookingDtoOutput getById(long bookingId, long userId) {
+        throwIfBookingNotFound(bookingId);
+        throwIfUserNotFound(userId);
+        if (bookingRepository.findById(bookingId).get().getBooker().getId() != userId) {
```
bookingrepository.findById(bookingId).get() 

Здесь для начала нужно проверить, есть ли этот объект, isPresent() а то может упасть ошибка

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/item/comment/CommentRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface CommentRepository extends JpaRepository<Comment, Long> {
+
+    List<Comment> getAllByItem_Id(long id);
```
Подчеркивание нужно убрать из названия метода

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -3,59 +3,134 @@
         throwIfUserNotFound(userId);
-        item.setOwner(userRepositoryInMemory.getById(userId));
-        return ItemMapper.toItemDto(itemRepositoryInMemory.create(item));
+        item.setOwner(userRepository.findById(userId).get());
```
Когда вы получаете объект по методу findById - там же возвращается Optional - здесь всегда нужно проверять на isPresent().
В остальных местах тоже сразу поменяйте. 
Можно создать отдельный метод 

```
private Item item(long id) {
        return itemRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Item with id not found " + id));
    }
```

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -3,19 +3,45 @@
         this.email = email;
     }
+
+    public User() {
+    }
```
Вы же вовсею используете ломбок, поставьте анотацию @NoArgsConstructor

NekhoroshkinaElena committed:
refactor: проставила аннотацию lombok-@value для некоторых ДТО классов
fix: убрать аннотацию @column для поля status в классе Booking
fix: добавить проверки при плучении объектов по методу findById()
fix: убрать подчеркивания из названия методов в репозиториях CommentRepository и BookingRepository
refactor: поставить аннотацию @NoArgsConstructor и убрать пустые конструкторы

NekhoroshkinaElena commented:
Большое спасибо за подсказку: 
"Можно создать отдельный метод

private Item item(long id) {
        return itemRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Item with id not found " + id));
    }
" я как раз не могла понять как реализовать такой метод проверки:)
